pygmt/src/tilemap.py

```diff
@@ -0,0 +1,136 @@
+        ``"auto"`` to automatically determine the zoom level based on the
+        bounding box region extent].
+
+        **Note**: The maximum possible zoom level may be smaller than ``22``,
+        and depends on what is supported by the chosen web tile provider
+        source.
+
+    source : xyzservices.TileProvider or str
+        Optional. The tile source: web tile provider or path to a local file.
+        Provide either:
+
+        - A web tile provider in the form of a
+          :class:`xyzservices.TileProvider` object. See
+          :doc:`Contextily providers <contextily:providers_deepdive>` for a
+          list of tile providers [Default is
+          ``xyzservices.providers.Stamen.Terrain``, i.e. Stamen Terrain web
+          tiles].
+        - A web tile provider in the form of a URL. The placeholders for the
+          XYZ in the URL need to be {{x}}, {{y}}, {{z}}, respectively. E.g.
+          ``https://{{s}}.tile.openstreetmap.org/{{z}}/{{x}}/{{y}}.png``.
+        - A local file path. The file is read with
+          :doc:`rasterio <rasterio:index>` and all bands are loaded into the
+          basemap. See
+          :doc:`contextily:working_with_local_files`.
+
+        IMPORTANT: Tiles are assumed to be in the Spherical Mercator projection
+        (EPSG:3857).
+
+    lonlat : bool
+        Optional. If ``False``, coordinates in ``region`` are assumed to be
+        Spherical Mercator as opposed to longitude/latitude [Default is
+        ``True``].
+
+    wait : int
+        Optional. If the tile API is rate-limited, the number of seconds to
+        wait between a failed request and the next try [Default is ``0``].
+
+    max_retries : int
+        Optional. Total number of rejected requests allowed before contextily
+        will stop trying to fetch more tiles from a rate-limited API [Default
+        is ``2``].
+
+    kwargs : dict
+        Extra keyword arguments to pass to :meth:`pygmt.Figure.grdimage`.
+
+    Raises
+    ------
+    ModuleNotFoundError
+        If ``rioxarray`` is not installed. Follow
+        :doc:`install instructions for rioxarray <rioxarray:installation>`,
+        (e.g. via ``pip install rioxarray``) before using this function.
+    """
+    kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
+
+    if rioxarray is None:
+        raise ModuleNotFoundError(
+            "Package `rioxarray` is required to be installed to use this function. "
+            "Please use `pip install rioxarray` or "
+            "`mamba install -c conda-forge rioxarray` "
+            "to install the package."
+        )
+
+    with GMTTempFile(suffix=".tif") as tmpfile:
+        raster = load_tile_map(
+            region=region,
+            zoom=zoom,
+            source=source,
+            lonlat=lonlat,
+            wait=wait,
+            max_retries=max_retries,
+        )
+        raster.rio.to_raster(raster_path=tmpfile.name)
```

Comment:
Move `load_tile_map` out of the `GMTTempFile` context:
```suggestion
    raster = load_tile_map(
        region=region,
        zoom=zoom,
        source=source,
        lonlat=lonlat,
        wait=wait,
        max_retries=max_retries,
    )

    with GMTTempFile(suffix=".tif") as tmpfile:
        raster.rio.to_raster(raster_path=tmpfile.name)
```